src/Compilers/Test/Core/Assert/AssertEx.cs

```diff
@@ -594,37 +616,117 @@ public static void AssertResultsEqual(string result1, string result2)
-            actualString = NormalizeWhitespace(actualString);
-            Assert.Contains(expectedSubString, actualString, StringComparison.Ordinal);
+            AssertContainsToleratingWhitespaceDifferences(expectedSubString.AsSpan(), actualString.AsSpan());
+        }
+
+        public static void AssertContainsToleratingWhitespaceDifferences(ReadOnlySpan<char> expectedSubString, ReadOnlySpan<char> actualString)
+        {
+            expectedSubString = NormalizeWhitespace(expectedSubString, out var pooled1);
+            actualString = NormalizeWhitespace(actualString, out var pooled2);
+
+            if (!actualString.Contains(expectedSubString, StringComparison.Ordinal))
+            {
+                // Call through to actual API to get same error message experience
+                Assert.Contains(expectedSubString.ToString(), actualString.ToString(), StringComparison.Ordinal);
+            }
+
+            if (pooled1 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled1);
+            }
+            if (pooled2 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled2);
+            }
         }
 
         public static void AssertStartsWithToleratingWhitespaceDifferences(string expectedSubString, string actualString)
         {
-            expectedSubString = NormalizeWhitespace(expectedSubString);
-            actualString = NormalizeWhitespace(actualString);
-            Assert.StartsWith(expectedSubString, actualString, StringComparison.Ordinal);
+            AssertStartsWithToleratingWhitespaceDifferences(expectedSubString.AsSpan(), actualString.AsSpan());
+        }
+
+        public static void AssertStartsWithToleratingWhitespaceDifferences(ReadOnlySpan<char> expectedSubString, ReadOnlySpan<char> actualString)
+        {
+            expectedSubString = NormalizeWhitespace(expectedSubString, out var pooled1);
+            actualString = NormalizeWhitespace(actualString, out var pooled2);
+
+            if (!actualString.StartsWith(expectedSubString, StringComparison.Ordinal))
+            {
+                // Call through to actual API to get same error message experience
+                Assert.StartsWith(expectedSubString.ToString(), actualString.ToString(), StringComparison.Ordinal);
+            }
+
+            if (pooled1 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled1);
+            }
+            if (pooled2 != null)
+            {
+                ArrayPool<char>.Shared.Return(pooled2);
+            }
+        }
+
+        internal static ReadOnlySpan<char> NormalizeWhitespace(ReadOnlySpan<char> input, out char[] pooledAlloc)
+        {
+            return NormalizeWhitespaceROM(input, out pooledAlloc).Span;
         }
 
-        internal static string NormalizeWhitespace(string input)
+        internal static ReadOnlyMemory<char> NormalizeWhitespaceROM(ReadOnlySpan<char> input, out char[] pooledAlloc)
```

Comment:
If we do suffix on the name we'd prefer the longer form vs. the abbreviation. Overall though I think I'd prefer we just have the one overload that returns `ReadOnlyMemory<char>` and then let the consumers do `.Span` as needed. That is a bit of ceremony but it's one that we've already done in a few other places.